packages/core-data/src/hooks/use-entity-records.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * Internal dependencies
+ */
+import useQuerySelect from './use-query-select';
+import { store as coreStore } from '../';
+import { Status } from './constants';
+
+interface EntityRecordsResolution< RecordType > {
+	/** The requested entity record */
+	records: RecordType[] | null;
+
+	/**
+	 * Is the record still being resolved?
+	 */
+	isResolving: boolean;
+
+	/**
+	 * Is the record resolved by now?
+	 */
+	hasResolved: boolean;
+
+	/** Resolution status */
+	status: Status;
+}
+
+/**
+ * Resolves the specified entity records.
+ *
+ * @param  kind      Kind of the requested entities.
+ * @param  name      Name of the requested entities.
+ * @param  queryArgs HTTP query for the requested entities.
+ *
+ * @example
+ * ```js
+ * import { useEntityRecord } from '@wordpress/core-data';
+ *
+ * function PageTitlesList() {
+ *   const { records, isResolving } = useEntityRecords( 'postType', 'page' );
+ *
+ *   if ( isResolving ) {
+ *     return 'Loading...';
+ *   }
+ *
+ *   return (
+ *     <ul>
+ *       {records.map(( page ) => (
+ *         <li>{ page.title }</li>
+ *       ))}
+ *     </ul>
+ *   );
+ * }
+ *
+ * // Rendered in the application:
+ * // <PageTitlesList />
+ * ```
+ *
+ * In the above example, when `PageTitlesList` is rendered into an
+ * application, the list of records and the resolution details will be retrieved from
+ * the store state using `getEntityRecords()`, or resolved if missing.
+ *
+ * @return {EntityRecordsResolution<RecordType>} Entity records data.
+ * @template RecordType
+ */
+export default function __experimentalUseEntityRecords< RecordType >(
+	kind: string,
+	name: string,
+	queryArgs: unknown = {}
```

Comment:
how hard would it be to surface a type of `queryArgs` here from wherever it's consumed? if not, can we add a link in the JSDoc comment to point people to where they can find out? I know it's passed-through to `getEntityRecords` but `unknown` isn't that helpful.

Given that these _are_ just passed through we might be able to reuse the types from `getEntityRecords`. I'm curious if the rename from `data` to `records` is fully worth it considering that it incurs a needless object clone, but I get how it's nice to match with `useEntityRecord`

```ts
interface EntityRecordHook<RecordType> {
    (kind: string; name: string; queryArgs: HttpQuery) => EntityRecordResolution<RecordType>;
}


// use-entity-record.ts
export default useEntityRecord: EntityRecordHook = (kind, name, queryArgs) => …;

// use-entity-records.ts
export default useEntityRecords: EntityRecordHook = …
```

https://user-images.githubusercontent.com/5431237/154117532-3f6bcc65-9b2d-4a2e-aa37-da16832faa92.mov

[playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgBJjABwIoFdoCeyA3gL4BQ5oksiKACgPYDOYJ5ynywAJgFzIQuALYAjaAG4OXTCzAB9BI3ARwA1lFABzKRSoraSZAFVmEPIQBKEZowA2uMMGUAeACoA+dl2Q84YOAE3KR8ACzhma1s7ADcIfmRRRnsIOBAQrmBIm3sY7QEklLTdSmpoeCMAUXBgMAJrJSgeKPtHZxAXBsYmtwJMCC8IAA9IEB5mZAB5YVqXU3N8KHqchydXLp6+gYAaZAByPwC9r2JpTigIRoSNnl7+kv0aCpRqpzqb1GSAa28uTsvurcth4ABRnHzIAD0ACpocgAML+CBabpERgwZBKFTgZBgRjIGAQMAIUK7CAAOi05OQACJZKw7hAaZxoZDwT4vqAEhptNt2VwYXCAMr9BDAGDABCY5SjNgAd1qoVAmKRKKWZMp1LpcmZ3VpmDgWiZyFZ-M4IDgwgg6jAmhAWj5EIFsIRuFYjGEaDcbnovhsCE0mDWIAJetCjDluPxhOJoWQFyu3AxMD1cFxW0EjDYmCgjDyPHiiSIjQpJrZTs4AEdFgQAIJQLTMAD8AnQWAsS3BAEoBK9assri1Vu1-ldGR4HsNZFA2FjWMg3RA++8AU1kABeEikZARZDLgeAz6ML5SchT7qz5Tzxf7m7MXs1FdXI8-TcgzljXYWq27auEeuNl2G4nOCFxgLgUAhjej4Hk075cl+loQL+NYAcwXYPHObDEAmgLbpu0FvLB4wuEwrCgns9JgIyey7JRch7Bh5C4U05JUYoMqqGw648fsWgXP4TZ7EAA)

Reply:
> how hard would it be to surface a type of queryArgs here from wherever it's consumed? if not, can we add a link in the JSDoc comment to point people to where they can find out? I know it's passed-through to getEntityRecords but unknown isn't that helpful. Given that these are just passed through we might be able to reuse the types from getEntityRecords. I

Good idea! The `EntityQuery` type from https://github.com/WordPress/gutenberg/pull/40024 would make a good fit here once that PR lands. 

> I'm curious if the rename from data to records is fully worth it considering that it incurs a needless object clone, but I get how it's nice to match with useEntityRecord

It doesn't match, though 😅  it is `record` and `records`. The reason I went for that versus `data` is for the sake of the consumer of this hook. Referring to a `record` inside of a component is clearer than referring to an opaque `data`. I am not a fan of that destructuring, though.